imsim/opd.py

```diff
@@ -0,0 +1,193 @@
+            specified in the (rotated) coordinate system of the telescope's
+            entrance pupil (usually the primary mirror).
+
+    Optional config inputs:
+        rotTelPos: Angle
+            Rotation applied to field angles.  Default: 0.0 radians.
+            This is useful for aligning field angles with particular detectors
+            when a camera rotator has been engaged.  If both this keyword and
+            the telescope's rotTelPos keyword are equal, then the fields sampled
+            will correspond to the same focal plane positions as when both
+            rotTelPos keywords are 0.0.
+        nx: int
+            Size of OPD image in pixels.  Default: 255.
+        wavelength: float
+            Wavelength of light in nanometers.  Default: None.
+            If not specified, then the wavelength will be taken from the
+            current bandpass.
+        projection: str
+            Projection mapping field angles to spherical coordinates.
+            Default: 'postel'.
+            See batoid documentation for more details.
+        sphereRadius: float
+            Radius of reference sphere in meters.  Default: None.
+            If not specified, then the radius will be taken from the telescope
+            object.
+        reference: {'chief', 'mean'}
+            Reference point for measuring OPD.  Default: 'chief'.
+            See batoid documentation for more details.
+        eps: float
+            Annular Zernike obscuration fraction.  Default: None.
+            If not specified, then the value will be taken from the telescope
+            object.
+        jmax: int
+            Maximum Zernike order.  Default: 28.
+
+    Notes:
+        The OPD image coordinates are always aligned with the entrance pupil,
+        regardless of the value of rotTelPos.  The OPD values are in nm, with
+        NaN values corresponding to vignetted regions.  The OPD is always
+        computed for the fiducial telescope focal plane height; i.e., it ignores
+        any detector-by-detector offsets in focal plane height.
+    """
+    def initialize(self, data, scratch, config, base, logger):
+        req = {
+            'file_name': str,
+            'fields': list
+        }
+        opt = {
+            'rotTelPos': galsim.Angle,
+            'nx': int,
+            'wavelength': float,
+            'projection': str,
+            'sphereRadius': float,
+            'reference': str,
+            'eps': float,
+            'jmax': int
+        }
+        kwargs, safe = GetAllParams(config, base, req=req, opt=opt)
+
+        # Handle defaults for optional kwargs
+        self.rotTelPos = kwargs.pop('rotTelPos', 0.0*galsim.radians)
+        self.nx = kwargs.pop('nx', 255)
+        self.projection = kwargs.pop('projection', 'postel')
+        self.reference = kwargs.pop('reference', 'chief')
+        telescope = GetInputObj(
+            'telescope',
+            config,
+            base,
+            'telescope'
```

Comment:
Oh, one other extremely minor thing I just noticed.  The last thing here should be 'opd', not 'telescope'.  It's only used for error messages, saying which thing needed the telescope input item if it fails to load.